lightly/api/patch.py

```diff
@@ -0,0 +1,131 @@
+
+    Args:
+        rest_client_cls:
+            Must be the class swagger_client.rest.RESTClientObject to patch.
+            Note: it must be the class itself, not an instance of it.
+
+    Returns:
+
+    """
+    request = rest_client_cls.request
+
+    def request_patched(
+        self,
+        method,
+        url,
+        query_params=None,
+        headers=None,
+        body=None,
+        post_params=None,
+        _preload_content=True,
+        _request_timeout=None,
+    ):
+        if query_params is not None:
+            new_query_params = []
+            for name, value in query_params:
+                if isinstance(value, list):
+                    new_query_params.extend([(name, val) for val in value])
+                else:
+                    new_query_params.append((name, value))
+            query_params = new_query_params
+        return request(
+            method=method,
+            url=url,
+            query_params=query_params,
+            headers=headers,
+            body=body,
+            post_params=post_params,
+            _preload_content=_preload_content,
+            _request_timeout=_request_timeout,
+        )
+
+    return request_patched
+
+
+def make_swagger_generated_classes_picklable(
+    api_client_cls: Type,
+    configuration_cls: Type,
+    rest_client_cls: Type,
+) -> None:
+    """Adds __getstate__ and __setstate__ methods to swagger generated classes to make
+    them picklable."""
+    api_client_cls.__setstate__ = _ApiClient__setstate__
+    api_client_cls.__getstate__ = _ApiClient__getstate__
+    configuration_cls.__getstate__ = _Configuration__getstate__
+    configuration_cls.__setstate__ = _Configuration__setstate__
+    # For the rest client we only need __getstate__ as __setstate__ is handled by the
+    # api client.
+    rest_client_cls.__getstate__ = _RESTClientObject__getstate__
+
+
+def _Configuration__getstate__(self: Type) -> Dict[str, Any]:
+    state = self.__dict__.copy()
+    # Remove unpicklable entries.
+    state["logger"] = {}
+    state["logger_formatter"] = None
+    state["logger_stream_handler"] = None
+    state["logger_file_handler"] = None
+    return state
+
+
+def _Configuration__setstate__(self: Type, state: Dict[str, Any]) -> None:
+    self.__dict__.update(state)
+    # Recreate logger objects.
+    self.logger["package_logger"] = logging.getLogger(
+        "lightly.openapi_generated.swagger_client"
+    )
+    self.logger["urllib3_logger"] = logging.getLogger("urllib3")
+    # Set logger_format and logger_file explicitly because they recreate
+    # logger_formatter, logger_file_handler, and logger_stream_handler which are removed
+    # before pickling.
+    self.logger_format = state["_Configuration__logger_format"]
+    self.logger_file = state["_Configuration__logger_file"]
+    # Set debug explicitly because it has side effects.
+    self.debug = state["_Configuration__debug"]
```

Comment:
How did you figure out what you need to unset and what to reset? Asking because the fields in `__getstate__` and `__setstate__`.

Reply:
What to unset: Try to pickle a class, if it fails remove the offending attribute.

What to reset: Everything that is removed must be set again (with some exceptions). Make sure that `new_object = unpickle(pickle(object))` has the same attributes as original object.